File: DSA_Problems/Graphs/hasPath.js

const hasPath = (graph, src, dst) => {
    // DFS using Recursion
    if (src === dst) return true;

    for (let neighbor of graph[src]) {
         if (hasPath(graph, neighbor, dst)) {
            return true;
         }
    }

    return false;
}



const hasPath_bfs = (graph, src, dst) => {
    // Iterative, cannot use recursion
    const queue = [ src ];

    while (queue.length > 0) {
        const current = queue.shift();  // move the front of queue

        if (current === dst) return true;

        for (let neighbor of graph[current]) {
            queue.push(neighbor)
        }
    }

    return false;
}










const graph = {
    f: ['g', 'i'],
    g: ['h'],
    h: [],
    i: ['g', 'k'],
    j: ['i'],
    k: []
};


console.log(hasPath(graph, 'f', 'k'));  // true
console.log(hasPath_bfs(graph, 'f', 'k'));  // true

File: DSA_Problems/Graphs/hasPath.py


def has_path_dfs(graph, src, dst):
    if src == dst:
        return True

    for neighbor in graph[src]:
        if has_path_dfs(graph, neighbor, dst):
            return True
    return False


def has_path_bfs(graph, src, dst):
    queue = [src]

    while len(queue) > 0:
        current = queue.pop()
        if current == dst:
            return True
        for neighbor in graph[current]:
            queue.insert(0, neighbor)
    return False


graph = {
    'f': ['g', 'i'],
    'g': ['h'],
    'h': [],
    'i': ['g', 'k'],
    'j': ['i'],
    'k': []
}

if __name__ == '__main__':
    #print(has_path_dfs(graph, 'g', 'i'))
    print(has_path_bfs(graph, 'f', 'k'))
